File: libs/wasm-sdk/wasm-sdk-cpp/src/sdk.cpp
// Flow-Like WASM SDK – Implementation

#include "flow_like_sdk.h"
#include "../src/json.h"

namespace flowlike {

// ============================================================================
// NodeScores
// ============================================================================

std::string NodeScores::to_json() const {
    return "{\"privacy\":" + std::to_string(privacy) +
           ",\"security\":" + std::to_string(security) +
           ",\"performance\":" + std::to_string(performance) +
           ",\"governance\":" + std::to_string(governance) +
           ",\"reliability\":" + std::to_string(reliability) +
           ",\"cost\":" + std::to_string(cost) + "}";
}

// ============================================================================
// PinDefinition
// ============================================================================

PinDefinition PinDefinition::input(const std::string& name,
                                   const std::string& friendly_name,
                                   const std::string& description,
                                   DataType data_type) {
    PinDefinition p;
    p.name          = name;
    p.friendly_name = friendly_name;
    p.description   = description;
    p.pin_type      = PinType::Input;
    p.data_type     = data_type;
    return p;
}

PinDefinition PinDefinition::output(const std::string& name,
                                    const std::string& friendly_name,
                                    const std::string& description,
                                    DataType data_type) {
    PinDefinition p;
    p.name          = name;
    p.friendly_name = friendly_name;
    p.description   = description;
    p.pin_type      = PinType::Output;
    p.data_type     = data_type;
    return p;
}

std::string PinDefinition::to_json() const {
    std::string j = "{\"name\":" + json::quote(name)
        + ",\"friendly_name\":" + json::quote(friendly_name)
        + ",\"description\":" + json::quote(description)
        + ",\"pin_type\":\"" + (pin_type == PinType::Input ? "Input" : "Output") + "\""
        + ",\"data_type\":\"" + data_type_str(data_type) + "\"";
    if (!default_value.empty()) j += ",\"default_value\":" + default_value;
    if (!value_type.empty())    j += ",\"value_type\":" + json::quote(value_type);
    if (!schema.empty())        j += ",\"schema\":" + json::quote(schema);
    j += "}";
    return j;
}

// ============================================================================
// NodeDefinition
// ============================================================================

std::string NodeDefinition::to_json() const {
    std::string pins_json = "[";
    for (size_t i = 0; i < pins.size(); ++i) {
        if (i > 0) pins_json += ",";
        pins_json += pins[i].to_json();
    }
    pins_json += "]";

    std::string j = "{\"name\":" + json::quote(name)
        + ",\"friendly_name\":" + json::quote(friendly_name)
        + ",\"description\":" + json::quote(description)
        + ",\"category\":" + json::quote(category)
        + ",\"pins\":" + pins_json
        + ",\"long_running\":" + (long_running ? "true" : "false")
        + ",\"abi_version\":" + std::to_string(abi_version);
    if (!icon.empty())  j += ",\"icon\":" + json::quote(icon);
    if (has_scores)     j += ",\"scores\":" + scores.to_json();
    if (!docs.empty())  j += ",\"docs\":" + json::quote(docs);
    if (!permissions.empty()) {
        j += ",\"permissions\":[";
        for (size_t i = 0; i < permissions.size(); ++i) {
            if (i > 0) j += ",";
            j += json::quote(permissions[i]);
        }
        j += "]";
    }
    j += "}";
    return j;
}

// ============================================================================
// ExecutionResult
// ============================================================================

std::string ExecutionResult::to_json() const {
    std::string out_json = "{";
    bool first = true;
    for (const auto& kv : outputs) {
        if (!first) out_json += ",";
        out_json += json::quote(kv.first) + ":" + kv.second;
        first = false;
    }
    out_json += "}";

    std::string exec_json = "[";
    for (size_t i = 0; i < activate_exec.size(); ++i) {
        if (i > 0) exec_json += ",";
        exec_json += json::quote(activate_exec[i]);
    }
    exec_json += "]";

    std::string j = "{\"outputs\":" + out_json
        + ",\"activate_exec\":" + exec_json
        + ",\"pending\":" + (pending ? "true" : "false");
    if (!error.empty()) j += ",\"error\":" + json::quote(error);
    j += "}";
    return j;
}

// ============================================================================
// Parse execution input
// ============================================================================

ExecutionInput parse_execution_input(const std::string& raw) {
    ExecutionInput inp;
    json::parse_inputs(raw, inp.inputs);
    inp.node_id      = json::extract_string(raw, "node_id");
    inp.node_name    = json::extract_string(raw, "node_name");
    inp.run_id       = json::extract_string(raw, "run_id");
    inp.app_id       = json::extract_string(raw, "app_id");
    inp.board_id     = json::extract_string(raw, "board_id");
    inp.user_id      = json::extract_string(raw, "user_id");
    inp.stream_state = json::extract_bool(raw, "stream_state");
    inp.log_level    = static_cast<uint8_t>(json::extract_int(raw, "log_level"));
    return inp;
}

// ============================================================================
// Serialization
// ============================================================================

int64_t serialize_definition(const NodeDefinition& def) {
    return pack_result(def.to_json());
}

int64_t serialize_result(const ExecutionResult& result) {
    return pack_result(result.to_json());
}

}  // namespace flowlike

File: templates/wasm-node-python/src/sdk.py
"""
Flow-Like WASM SDK for Python

Provides types, utilities, and base classes for building WASM nodes
that can be executed by the Flow-Like runtime.

Usage:
    from sdk import NodeDefinition, PinDefinition, Context, ExecutionResult

    def get_definition() -> NodeDefinition:
        node = NodeDefinition("my_node", "My Node", "Does something", "Custom/Python")
        node.add_pin(PinDefinition.input_exec("exec"))
        node.add_pin(PinDefinition.input_pin("text", "String", default=""))
        node.add_pin(PinDefinition.output_exec("exec_out"))
        node.add_pin(PinDefinition.output_pin("result", "String"))
        return node

    def run(ctx: Context) -> ExecutionResult:
        text = ctx.get_string("text") or ""
        ctx.set_output("result", text.upper())
        return ctx.success()
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Any


ABI_VERSION = 1


class LogLevel(IntEnum):
    DEBUG = 0
    INFO = 1
    WARN = 2
    ERROR = 3
    FATAL = 4


class PinType:
    EXEC = "Exec"
    STRING = "String"
    I64 = "I64"
    F64 = "F64"
    BOOL = "Bool"
    GENERIC = "Generic"
    BYTES = "Bytes"

    _ALL = {EXEC, STRING, I64, F64, BOOL, GENERIC, BYTES}

    @classmethod
    def validate(cls, data_type: str) -> str:
        if data_type not in cls._ALL:
            raise ValueError(f"Invalid pin data type: {data_type}. Must be one of {cls._ALL}")
        return data_type


def _humanize(name: str) -> str:
    return " ".join(w.capitalize() for w in name.split("_") if w)


@dataclass
class NodeScores:
    privacy: int = 0
    security: int = 0
    performance: int = 0
    governance: int = 0
    reliability: int = 0
    cost: int = 0

    def to_dict(self) -> dict[str, int]:
        return {
            "privacy": self.privacy,
            "security": self.security,
            "performance": self.performance,
            "governance": self.governance,
            "reliability": self.reliability,
            "cost": self.cost,
        }


@dataclass
class PinDefinition:
    name: str
    friendly_name: str
    description: str
    pin_type: str  # "Input" or "Output"
    data_type: str
    default_value: Any = None
    value_type: str | None = None
    schema: str | None = None
    valid_values: list[str] | None = None
    range: tuple[float, float] | None = None

    @classmethod
    def input_pin(
        cls,
        name: str,
        data_type: str,
        *,
        description: str = "",
        default: Any = None,
        friendly_name: str | None = None,
    ) -> PinDefinition:
        PinType.validate(data_type)
        return cls(
            name=name,
            friendly_name=friendly_name or _humanize(name),
            description=description or f"Input: {name}",
            pin_type="Input",
            data_type=data_type,
            default_value=default,
        )

    @classmethod
    def output_pin(
        cls,
        name: str,
        data_type: str,
        *,
        description: str = "",
        friendly_name: str | None = None,
    ) -> PinDefinition:
        PinType.validate(data_type)
        return cls(
            name=name,
            friendly_name=friendly_name or _humanize(name),
            description=description or f"Output: {name}",
            pin_type="Output",
            data_type=data_type,
        )

    @classmethod
    def input_exec(cls, name: str = "exec", *, description: str = "") -> PinDefinition:
        return cls(
            name=name,
            friendly_name=_humanize(name),
            description=description or f"Input: {name}",
            pin_type="Input",
            data_type=PinType.EXEC,
        )

    @classmethod
    def output_exec(cls, name: str = "exec_out", *, description: str = "") -> PinDefinition:
        return cls(
            name=name,
            friendly_name=_humanize(name),
            description=description or f"Output: {name}",
            pin_type="Output",
            data_type=PinType.EXEC,
        )

    def with_default(self, value: Any) -> PinDefinition:
        self.default_value = value
        return self

    def with_value_type(self, value_type: str) -> PinDefinition:
        self.value_type = value_type
        return self

    def with_schema(self, schema: str) -> PinDefinition:
        self.schema = schema
        return self

    def with_valid_values(self, values: list[str]) -> PinDefinition:
        self.valid_values = values
        return self

    def with_range(self, min_val: float, max_val: float) -> PinDefinition:
        self.range = (min_val, max_val)
        return self

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "name": self.name,
            "friendly_name": self.friendly_name,
            "description": self.description,
            "pin_type": self.pin_type,
            "data_type": self.data_type,
        }
        if self.default_value is not None:
            d["default_value"] = self.default_value
        if self.value_type is not None:
            d["value_type"] = self.value_type
        if self.schema is not None:
            d["schema"] = self.schema
        if self.valid_values is not None:
            d["valid_values"] = self.valid_values
        if self.range is not None:
            d["range"] = list(self.range)
        return d


@dataclass
class NodeDefinition:
    name: str
    friendly_name: str
    description: str
    category: str
    icon: str | None = None
    pins: list[PinDefinition] = field(default_factory=list)
    scores: NodeScores | None = None
    long_running: bool | None = None
    docs: str | None = None
    permissions: list[str] = field(default_factory=list)
    abi_version: int = ABI_VERSION

    def add_pin(self, pin: PinDefinition) -> NodeDefinition:
        self.pins.append(pin)
        return self

    def set_scores(self, scores: NodeScores) -> NodeDefinition:
        self.scores = scores
        return self

    def set_long_running(self, long_running: bool) -> NodeDefinition:
        self.long_running = long_running
        return self

    def add_permission(self, permission: str) -> NodeDefinition:
        self.permissions.append(permission)
        return self

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "name": self.name,
            "friendly_name": self.friendly_name,
            "description": self.description,
            "category": self.category,
            "pins": [p.to_dict() for p in self.pins],
            "abi_version": self.abi_version,
        }
        if self.icon is not None:
            d["icon"] = self.icon
        if self.scores is not None:
            d["scores"] = self.scores.to_dict()
        if self.long_running is not None:
            d["long_running"] = self.long_running
        if self.docs is not None:
            d["docs"] = self.docs
        if self.permissions:
            d["permissions"] = self.permissions
        return d

    def to_json(self) -> str:
        return json.dumps(self.to_dict())


@dataclass
class ExecutionInput:
    inputs: dict[str, Any]
    node_id: str = ""
    run_id: str = ""
    app_id: str = ""
    board_id: str = ""
    user_id: str = ""
    stream_state: bool = False
    log_level: int = LogLevel.INFO
    node_name: str = ""

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> ExecutionInput:
        return cls(
            inputs=data.get("inputs", {}),
            node_id=data.get("node_id", ""),
            run_id=data.get("run_id", ""),
            app_id=data.get("app_id", ""),
            board_id=data.get("board_id", ""),
            user_id=data.get("user_id", ""),
            stream_state=data.get("stream_state", False),
            log_level=data.get("log_level", LogLevel.INFO),
            node_name=data.get("node_name", ""),
        )

    @classmethod
    def from_json(cls, json_str: str) -> ExecutionInput:
        return cls.from_dict(json.loads(json_str))


@dataclass
class ExecutionResult:
    outputs: dict[str, Any] = field(default_factory=dict)
    error: str | None = None
    activate_exec: list[str] = field(default_factory=list)
    pending: bool | None = None

    @classmethod
    def ok(cls) -> ExecutionResult:
        return cls()

    @classmethod
    def fail(cls, message: str) -> ExecutionResult:
        return cls(error=message)

    def set_output(self, name: str, value: Any) -> ExecutionResult:
        self.outputs[name] = value
        return self

    def exec(self, pin_name: str) -> ExecutionResult:
        self.activate_exec.append(pin_name)
        return self

    def set_pending(self, pending: bool) -> ExecutionResult:
        self.pending = pending
        return self

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "outputs": self.outputs,
            "activate_exec": self.activate_exec,
        }
        if self.error is not None:
            d["error"] = self.error
        if self.pending is not None:
            d["pending"] = self.pending
        return d

    def to_json(self) -> str:
        return json.dumps(self.to_dict())


class HostBridge:
    """Interface for host function calls. Replaced at runtime by actual WASM host imports."""

    def log(self, level: int, message: str) -> None:
        pass

    def stream(self, event_type: str, data: str) -> None:
        pass

    def get_variable(self, name: str) -> Any:
        return None

    def set_variable(self, name: str, value: Any) -> bool:
        return False

    def time_now(self) -> int:
        return 0

    def random(self) -> int:
        return 0


class MockHostBridge(HostBridge):
    """Host bridge for local testing with captured logs and streams."""

    def __init__(self) -> None:
        self.logs: list[tuple[int, str]] = []
        self.streams: list[tuple[str, str]] = []
        self.variables: dict[str, Any] = {}
        self._time: int = 0
        self._random_value: int = 42

    def log(self, level: int, message: str) -> None:
        self.logs.append((level, message))

    def stream(self, event_type: str, data: str) -> None:
        self.streams.append((event_type, data))

    def get_variable(self, name: str) -> Any:
        return self.variables.get(name)

    def set_variable(self, name: str, value: Any) -> bool:
        self.variables[name] = value
        return True

    def time_now(self) -> int:
        return self._time

    def random(self) -> int:
        return self._random_value


# Global host bridge (swapped for testing or WASM runtime)
_host: HostBridge = HostBridge()


def set_host(host: HostBridge) -> None:
    global _host
    _host = host


def get_host() -> HostBridge:
    return _host


class Context:
    """Execution context providing typed input access, output setting, logging, and streaming."""

    def __init__(self, execution_input: ExecutionInput, host: HostBridge | None = None) -> None:
        self._input = execution_input
        self._result = ExecutionResult.ok()
        self._host = host or _host

    @classmethod
    def from_dict(cls, data: dict[str, Any], host: HostBridge | None = None) -> Context:
        return cls(ExecutionInput.from_dict(data), host)

    @classmethod
    def from_json(cls, json_str: str, host: HostBridge | None = None) -> Context:
        return cls(ExecutionInput.from_json(json_str), host)

    # -- Metadata --

    @property
    def node_id(self) -> str:
        return self._input.node_id

    @property
    def node_name(self) -> str:
        return self._input.node_name

    @property
    def run_id(self) -> str:
        return self._input.run_id

    @property
    def app_id(self) -> str:
        return self._input.app_id

    @property
    def board_id(self) -> str:
        return self._input.board_id

    @property
    def user_id(self) -> str:
        return self._input.user_id

    @property
    def stream_enabled(self) -> bool:
        return self._input.stream_state

    @property
    def log_level(self) -> int:
        return self._input.log_level

    # -- Input getters --

    def get_input(self, name: str) -> Any:
        return self._input.inputs.get(name)

    def get_string(self, name: str, default: str | None = None) -> str | None:
        val = self.get_input(name)
        if val is None:
            return default
        return str(val)

    def get_i64(self, name: str, default: int | None = None) -> int | None:
        val = self.get_input(name)
        if val is None:
            return default
        return int(val)

    def get_f64(self, name: str, default: float | None = None) -> float | None:
        val = self.get_input(name)
        if val is None:
            return default
        return float(val)

    def get_bool(self, name: str, default: bool | None = None) -> bool | None:
        val = self.get_input(name)
        if val is None:
            return default
        return bool(val)

    def require_input(self, name: str) -> Any:
        val = self.get_input(name)
        if val is None:
            raise ValueError(f"Required input '{name}' not provided")
        return val

    # -- Output setters --

    def set_output(self, name: str, value: Any) -> None:
        self._result.set_output(name, value)

    def activate_exec(self, pin_name: str) -> None:
        self._result.exec(pin_name)

    def set_pending(self, pending: bool) -> None:
        self._result.set_pending(pending)

    # -- Logging (level-gated) --

    def debug(self, message: str) -> None:
        if self._input.log_level <= LogLevel.DEBUG:
            self._host.log(LogLevel.DEBUG, message)

    def info(self, message: str) -> None:
        if self._input.log_level <= LogLevel.INFO:
            self._host.log(LogLevel.INFO, message)

    def warn(self, message: str) -> None:
        if self._input.log_level <= LogLevel.WARN:
            self._host.log(LogLevel.WARN, message)

    def error(self, message: str) -> None:
        if self._input.log_level <= LogLevel.ERROR:
            self._host.log(LogLevel.ERROR, message)

    # -- Streaming --

    def stream_text(self, text: str) -> None:
        if self._input.stream_state:
            self._host.stream("text", text)

    def stream_json(self, data: Any) -> None:
        if self._input.stream_state:
            self._host.stream("json", json.dumps(data))

    def stream_progress(self, progress: float, message: str) -> None:
        if self._input.stream_state:
            payload = json.dumps({"progress": progress, "message": message})
            self._host.stream("progress", payload)

    # -- Variables --

    def get_variable(self, name: str) -> Any:
        return self._host.get_variable(name)

    def set_variable(self, name: str, value: Any) -> bool:
        return self._host.set_variable(name, value)

    # -- Finalize --

    def success(self) -> ExecutionResult:
        self._result.exec("exec_out")
        return self._result

    def fail(self, error: str) -> ExecutionResult:
        self._result.error = error
        return self._result

    def finish(self) -> ExecutionResult:
        return self._result


def node(
    name: str,
    friendly_name: str,
    description: str,
    category: str,
    **kwargs: Any,
) -> NodeDefinition:
    """Shorthand to create a NodeDefinition."""
    nd = NodeDefinition(name, friendly_name, description, category, **kwargs)
    return nd
